File: Repo/opencog/cogutil/Go/opencog-cogutil.go
/*
 * opencog-cogutil.go
 * 
 * OpenCog Cogutil - Go Utility Library
 * A collection of utility functions and classes for OpenCog framework
 * 
 * This single-file implementation demonstrates Go's strengths:
 * - Simplicity and readability
 * - Built-in concurrency with goroutines and channels
 * - Struct embedding for composition
 * - Interfaces for polymorphism
 * - Fast compilation and execution
 */

package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
	"strings"
	"time"
	"unicode"
)

// ===== Logger System =====
// Demonstrates: Enums with iota, methods on structs

type LogLevel int

const (
	DEBUG LogLevel = iota
	INFO
	WARN
	ERROR
)

func (l LogLevel) String() string {
	switch l {
	case DEBUG:
		return "DEBUG"
	case INFO:
		return "INFO"
	case WARN:
		return "WARN"
	case ERROR:
		return "ERROR"
	default:
		return "UNKNOWN"
	}
}

type Logger struct {
	name     string
	minLevel LogLevel
}

func NewLogger(name string, level LogLevel) *Logger {
	return &Logger{
		name:     name,
		minLevel: level,
	}
}

func (l *Logger) log(level LogLevel, message string) {
	if level >= l.minLevel {
		timestamp := time.Now().Format("15:04:05")
		fmt.Printf("[%s] %s: %s\n", timestamp, level, message)
	}
}

func (l *Logger) Debug(message string) {
	l.log(DEBUG, message)
}

func (l *Logger) Info(message string) {
	l.log(INFO, message)
}

func (l *Logger) Warn(message string) {
	l.log(WARN, message)
}

func (l *Logger) Error(message string) {
	l.log(ERROR, message)
}

func (l *Logger) SetLevel(level LogLevel) {
	l.minLevel = level
}

// ===== Configuration Manager =====
// Demonstrates: Maps, error handling

type Config struct {
	data map[string]string
}

func NewConfig() *Config {
	return &Config{
		data: make(map[string]string),
	}
}

func (c *Config) Set(key, value string) {
	c.data[key] = value
}

func (c *Config) Get(key string) (string, bool) {
	value, ok := c.data[key]
	return value, ok
}

func (c *Config) GetOr(key, defaultValue string) string {
	if value, ok := c.data[key]; ok {
		return value
	}
	return defaultValue
}

func (c *Config) Has(key string) bool {
	_, ok := c.data[key]
	return ok
}

func (c *Config) LoadFromFile(filename string) error {
	file, err := os.Open(filename)
	if err != nil {
		return err
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		
		// Skip comments and empty lines
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}

		// Parse key=value pairs
		parts := strings.SplitN(line, "=", 2)
		if len(parts) == 2 {
			key := strings.TrimSpace(parts[0])
			value := strings.TrimSpace(parts[1])
			c.Set(key, value)
		}
	}

	return scanner.Err()
}

func (c *Config) Dump() {
	// Sort keys for consistent output
	keys := make([]string, 0, len(c.data))
	for k := range c.data {
		keys = append(keys, k)
	}
	sort.Strings(keys)

	for _, k := range keys {
		fmt.Printf("%s = %s\n", k, c.data[k])
	}
}

// ===== Timer Utility =====
// Demonstrates: Embedded time handling, methods

type Timer struct {
	name      string
	startTime time.Time
	elapsed   time.Duration
}

func NewTimer(name string) *Timer {
	return &Timer{
		name: name,
	}
}

func (t *Timer) Start() {
	t.startTime = time.Now()
}

func (t *Timer) Stop() time.Duration {
	t.elapsed = time.Since(t.startTime)
	return t.elapsed
}

func (t *Timer) Elapsed() time.Duration {
	return t.elapsed
}

func (t *Timer) ElapsedSeconds() float64 {
	return t.elapsed.Seconds()
}

// ===== String Utilities =====
// Demonstrates: String manipulation, slices

type StringUtils struct{}

func (StringUtils) Split(text string, delimiter rune) []string {
	parts := strings.FieldsFunc(text, func(r rune) bool {
		return r == delimiter
	})
	
	// Trim whitespace
	result := make([]string, 0, len(parts))
	for _, part := range parts {
		trimmed := strings.TrimSpace(part)
		if trimmed != "" {
			result = append(result, trimmed)
		}
	}
	return result
}

func (StringUtils) Join(strs []string, delimiter string) string {
	return strings.Join(strs, delimiter)
}

func (StringUtils) ToLower(text string) string {
	return strings.ToLower(text)
}

func (StringUtils) ToUpper(text string) string {
	return strings.ToUpper(text)
}

func (StringUtils) Trim(text string) string {
	return strings.TrimSpace(text)
}

func (StringUtils) CamelToSnake(text string) string {
	var result strings.Builder
	for i, r := range text {
		if unicode.IsUpper(r) && i > 0 {
			result.WriteRune('_')
		}
		result.WriteRune(unicode.ToLower(r))
	}
	return result.String()
}

func (StringUtils) SnakeToCamel(text string) string {
	parts := strings.Split(text, "_")
	var result strings.Builder
	
	for i, part := range parts {
		if i == 0 {
			result.WriteString(part)
		} else if len(part) > 0 {
			// Capitalize first letter manually (strings.Title is deprecated in Go 1.18+)
			result.WriteString(strings.ToUpper(string(part[0])) + part[1:])
		}
	}
	return result.String()
}

// ===== Performance Measurement =====
// Demonstrates: Higher-order functions, closures

func MeasurePerformance(operation string, f func()) {
	fmt.Printf("[PERF] Starting: %s\n", operation)
	start := time.Now()
	f()
	elapsed := time.Since(start)
	fmt.Printf("[PERF] Completed: %s in %.6fs\n", operation, elapsed.Seconds())
}

func TimeIt(name string, f func()) {
	start := time.Now()
	f()
	elapsed := time.Since(start)
	fmt.Printf("[TIMER] %s took %.6f seconds\n", name, elapsed.Seconds())
}

// ===== Concurrent Processing =====
// Demonstrates: Goroutines, channels, concurrency

func ProcessConcurrently(items []int, workers int) []int {
	jobs := make(chan int, len(items))
	results := make(chan int, len(items))

	// Start workers
	for w := 1; w <= workers; w++ {
		go func(id int) {
			for item := range jobs {
				// Simulate processing
				result := item * item
				results <- result
			}
		}(w)
	}

	// Send jobs
	for _, item := range items {
		jobs <- item
	}
	close(jobs)

	// Collect results
	processed := make([]int, 0, len(items))
	for i := 0; i < len(items); i++ {
		processed = append(processed, <-results)
	}

	return processed
}

// ===== Demonstration Function =====

func demonstrateCogutil() {
	fmt.Println("======================================================================")
	fmt.Println("OpenCog Cogutil - Go Utility Library Demo")
	fmt.Println("Showcasing Go's strengths: Simplicity, concurrency, fast compilation")
	fmt.Println("======================================================================")
	fmt.Println()

	// 1. Logger demonstration
	fmt.Println("1. Logger Demonstration")
	fmt.Println("--------------------------------------------------")
	logger := NewLogger("CogUtil", INFO)
	logger.Info("Cogutil library initialized")
	logger.Debug("This debug message won't show (level too low)")
	logger.Warn("This is a warning message")
	logger.Error("This is an error message")
	
	logger.SetLevel(DEBUG)
	logger.Debug("Now debug messages are visible")
	fmt.Println()

	// 2. Config demonstration
	fmt.Println("2. Configuration Manager")
	fmt.Println("--------------------------------------------------")
	config := NewConfig()
	config.Set("opencog.version", "1.0.0")
	config.Set("atomspace.enabled", "true")
	config.Set("cogserver.port", "17001")
	
	logger.Info("Configuration loaded:")
	config.Dump()
	fmt.Println()
	
	logger.Info(fmt.Sprintf("Port setting: %s", config.GetOr("cogserver.port", "unknown")))
	fmt.Println()

	// 3. Timer demonstration
	fmt.Println("3. Timer Utility")
	fmt.Println("--------------------------------------------------")
	timer := NewTimer("Processing")
	timer.Start()
	logger.Info("Simulating some work...")
	
	// Simulate work
	sum := 0
	for i := 0; i < 1000000; i++ {
		sum += i
	}
	_ = sum
	
	timer.Stop()
	logger.Info(fmt.Sprintf("Timer elapsed: %.6fs", timer.ElapsedSeconds()))
	fmt.Println()

	// 4. Performance measurement
	fmt.Println("4. Performance Measurement")
	fmt.Println("--------------------------------------------------")
	MeasurePerformance("Data processing", func() {
		time.Sleep(10 * time.Millisecond)
	})
	fmt.Println()

	// 5. String utilities
	fmt.Println("5. String Utilities")
	fmt.Println("--------------------------------------------------")
	logger.Info("String utilities demonstration:")
	
	utils := StringUtils{}
	text := "OpenCog,AtomSpace,CogServer,Cogutil"
	parts := utils.Split(text, ',')
	
	logger.Info("Split result:")
	for _, part := range parts {
		fmt.Printf("  - %s\n", part)
	}
	
	joined := utils.Join(parts, " + ")
	logger.Info(fmt.Sprintf("Joined: %s", joined))
	
	logger.Info(fmt.Sprintf("Uppercase: %s", utils.ToUpper("opencog rocks")))
	logger.Info(fmt.Sprintf("Lowercase: %s", utils.ToLower("OPENCOG ROCKS")))
	logger.Info(fmt.Sprintf("Trimmed: '%s'", utils.Trim("  spaced out  ")))
	
	// Go-specific: Case conversions
	logger.Info(fmt.Sprintf("camelCase → snake_case: %s", utils.CamelToSnake("myVariableName")))
	logger.Info(fmt.Sprintf("snake_case → camelCase: %s", utils.SnakeToCamel("my_variable_name")))
	fmt.Println()

	// 6. Function timing
	fmt.Println("6. Function Timing")
	fmt.Println("--------------------------------------------------")
	
	var result int
	TimeIt("fibonacci", func() {
		// Note: Using naive recursive fibonacci for demonstration
		// O(2^n) complexity - suitable for small n only
		fibonacci := func(n int) int {
			var fib func(int) int
			fib = func(n int) int {
				if n <= 1 {
					return n
				}
				return fib(n-1) + fib(n-2)
			}
			return fib(n)
		}
		result = fibonacci(25)
	})
	logger.Info(fmt.Sprintf("Fibonacci(25) = %d", result))
	fmt.Println()

	// 7. Slices and maps (Go-specific)
	fmt.Println("7. Slices and Maps (Go-specific)")
	fmt.Println("--------------------------------------------------")
	squares := make([]int, 10)
	for i := range squares {
		squares[i] = i * i
	}
	logger.Info(fmt.Sprintf("Squares: %v", squares))
	
	evens := make([]int, 0)
	for i := 0; i < 20; i++ {
		if i%2 == 0 {
			evens = append(evens, i)
		}
	}
	logger.Info(fmt.Sprintf("Even numbers: %v", evens))
	fmt.Println()

	// 8. Goroutines and channels (Go-specific)
	fmt.Println("8. Concurrency (Goroutines & Channels)")
	fmt.Println("--------------------------------------------------")
	logger.Info("Processing items concurrently with 4 workers...")
	
	items := []int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}
	processed := ProcessConcurrently(items, 4)
	
	logger.Info(fmt.Sprintf("Input: %v", items))
	logger.Info(fmt.Sprintf("Squared (concurrent): %v", processed))
	fmt.Println()

	// 9. Error handling
	fmt.Println("9. Error Handling (Go-specific)")
	fmt.Println("--------------------------------------------------")
	err := config.LoadFromFile("nonexistent.conf")
	if err != nil {
		logger.Warn(fmt.Sprintf("Config load failed (expected): %v", err))
	} else {
		logger.Info("Config loaded successfully")
	}
	fmt.Println()

	logger.Info("Cogutil demonstration complete!")
	fmt.Println("======================================================================")
	fmt.Println("Go strengths demonstrated:")
	fmt.Println("  ✓ Simple, readable syntax")
	fmt.Println("  ✓ Built-in concurrency (goroutines, channels)")
	fmt.Println("  ✓ Fast compilation and execution")
	fmt.Println("  ✓ Struct embedding for composition")
	fmt.Println("  ✓ Interfaces for polymorphism")
	fmt.Println("  ✓ Explicit error handling")
	fmt.Println("  ✓ Garbage collection")
	fmt.Println("======================================================================")
}

func main() {
	demonstrateCogutil()
}

File: Repo/opencog/cogutil/Rust/opencog-cogutil.rs
/*
 * opencog-cogutil.rs
 * 
 * OpenCog Cogutil - Rust Utility Library
 * A collection of utility functions and classes for OpenCog framework
 * 
 * This single-file implementation demonstrates Rust's strengths:
 * - Memory safety without garbage collection
 * - Ownership and borrowing for safe concurrency
 * - Zero-cost abstractions
 * - Pattern matching and Result/Option types
 * - Trait-based polymorphism
 */

use std::collections::HashMap;
use std::fs::File;
use std::io::{self, BufRead, BufReader, Write};
use std::time::{Duration, Instant, SystemTime, UNIX_EPOCH};
use std::fmt;

// ===== Logger System =====
// Demonstrates: Enums, pattern matching, trait implementation

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
enum LogLevel {
    Debug,
    Info,
    Warn,
    Error,
}

impl fmt::Display for LogLevel {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            LogLevel::Debug => write!(f, "DEBUG"),
            LogLevel::Info => write!(f, "INFO"),
            LogLevel::Warn => write!(f, "WARN"),
            LogLevel::Error => write!(f, "ERROR"),
        }
    }
}

struct Logger {
    name: String,
    min_level: LogLevel,
}

impl Logger {
    fn new(name: &str, level: LogLevel) -> Self {
        Logger {
            name: name.to_string(),
            min_level: level,
        }
    }

    fn log(&self, level: LogLevel, message: &str) {
        if level >= self.min_level {
            let timestamp = Self::get_timestamp();
            println!("[{}] {}: {}", timestamp, level, message);
        }
    }

    fn debug(&self, message: &str) {
        self.log(LogLevel::Debug, message);
    }

    fn info(&self, message: &str) {
        self.log(LogLevel::Info, message);
    }

    fn warn(&self, message: &str) {
        self.log(LogLevel::Warn, message);
    }

    fn error(&self, message: &str) {
        self.log(LogLevel::Error, message);
    }

    fn set_level(&mut self, level: LogLevel) {
        self.min_level = level;
    }

    fn get_timestamp() -> String {
        let now = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap();
        let secs = now.as_secs();
        let hours = (secs / 3600) % 24;
        let mins = (secs / 60) % 60;
        let secs = secs % 60;
        format!("{:02}:{:02}:{:02}", hours, mins, secs)
    }
}

// ===== Configuration Manager =====
// Demonstrates: HashMap, ownership, Result type

struct Config {
    data: HashMap<String, String>,
}

impl Config {
    fn new() -> Self {
        Config {
            data: HashMap::new(),
        }
    }

    fn set(&mut self, key: String, value: String) {
        self.data.insert(key, value);
    }

    fn get(&self, key: &str) -> Option<&String> {
        self.data.get(key)
    }

    fn get_or(&self, key: &str, default: &str) -> String {
        self.data.get(key).map(|s| s.clone()).unwrap_or_else(|| default.to_string())
    }

    fn has(&self, key: &str) -> bool {
        self.data.contains_key(key)
    }

    fn load_from_file(&mut self, filename: &str) -> Result<(), io::Error> {
        let file = File::open(filename)?;
        let reader = BufReader::new(file);

        for line in reader.lines() {
            let line = line?;
            let trimmed = line.trim();
            
            // Skip comments and empty lines
            if trimmed.is_empty() || trimmed.starts_with('#') {
                continue;
            }

            // Parse key=value pairs
            if let Some(pos) = trimmed.find('=') {
                let key = trimmed[..pos].trim().to_string();
                let value = trimmed[pos + 1..].trim().to_string();
                self.set(key, value);
            }
        }

        Ok(())
    }

    fn dump(&self) {
        let mut keys: Vec<_> = self.data.keys().collect();
        keys.sort();
        
        for key in keys {
            println!("{} = {}", key, self.data[key]);
        }
    }
}

// ===== Timer Utility =====
// Demonstrates: RAII pattern, Option type, method chaining

struct Timer {
    name: String,
    start_time: Option<Instant>,
    elapsed: Option<Duration>,
}

impl Timer {
    fn new(name: &str) -> Self {
        Timer {
            name: name.to_string(),
            start_time: None,
            elapsed: None,
        }
    }

    fn start(&mut self) {
        self.start_time = Some(Instant::now());
    }

    fn stop(&mut self) -> Duration {
        if let Some(start) = self.start_time {
            let elapsed = start.elapsed();
            self.elapsed = Some(elapsed);
            elapsed
        } else {
            Duration::from_secs(0)
        }
    }

    fn elapsed(&self) -> Option<Duration> {
        self.elapsed
    }

    fn elapsed_secs(&self) -> f64 {
        self.elapsed.map(|d| d.as_secs_f64()).unwrap_or(0.0)
    }
}

// ===== String Utilities =====
// Demonstrates: String handling, iterators, functional programming

struct StringUtils;

impl StringUtils {
    fn split(text: &str, delimiter: char) -> Vec<String> {
        text.split(delimiter)
            .map(|s| s.trim().to_string())
            .filter(|s| !s.is_empty())
            .collect()
    }

    fn join(strings: &[String], delimiter: &str) -> String {
        strings.join(delimiter)
    }

    fn to_lower(text: &str) -> String {
        text.to_lowercase()
    }

    fn to_upper(text: &str) -> String {
        text.to_uppercase()
    }

    fn trim(text: &str) -> String {
        text.trim().to_string()
    }

    fn camel_to_snake(text: &str) -> String {
        let mut result = String::new();
        for (i, c) in text.chars().enumerate() {
            if c.is_uppercase() && i > 0 {
                result.push('_');
            }
            result.push(c.to_lowercase().next().unwrap());
        }
        result
    }

    fn snake_to_camel(text: &str) -> String {
        let parts: Vec<&str> = text.split('_').collect();
        let mut result = String::new();
        
        for (i, part) in parts.iter().enumerate() {
            if i == 0 {
                result.push_str(part);
            } else {
                let mut chars = part.chars();
                if let Some(first) = chars.next() {
                    result.push(first.to_uppercase().next().unwrap());
                    result.push_str(&chars.collect::<String>());
                }
            }
        }
        result
    }
}

// ===== Performance Measurement =====
// Demonstrates: Closures, generic functions, higher-order functions

fn measure_performance<F, T>(operation: &str, f: F) -> T
where
    F: FnOnce() -> T,
{
    println!("[PERF] Starting: {}", operation);
    let start = Instant::now();
    let result = f();
    let elapsed = start.elapsed();
    println!("[PERF] Completed: {} in {:.6}s", operation, elapsed.as_secs_f64());
    result
}

// ===== Timing Decorator =====
// Demonstrates: Function wrapping, measurement

fn timeit<F, T>(name: &str, f: F) -> T
where
    F: FnOnce() -> T,
{
    let start = Instant::now();
    let result = f();
    let elapsed = start.elapsed();
    println!("[TIMER] {} took {:.6} seconds", name, elapsed.as_secs_f64());
    result
}

// ===== Demonstration Function =====

fn demonstrate_cogutil() {
    println!("======================================================================");
    println!("OpenCog Cogutil - Rust Utility Library Demo");
    println!("Showcasing Rust's strengths: Memory safety, zero-cost, ownership");
    println!("======================================================================");
    println!();

    // 1. Logger demonstration
    println!("1. Logger Demonstration");
    println!("--------------------------------------------------");
    let logger = Logger::new("CogUtil", LogLevel::Info);
    logger.info("Cogutil library initialized");
    logger.debug("This debug message won't show (level too low)");
    logger.warn("This is a warning message");
    logger.error("This is an error message");
    
    let mut logger_debug = Logger::new("CogUtil", LogLevel::Debug);
    logger_debug.set_level(LogLevel::Debug);
    logger_debug.debug("Now debug messages are visible");
    println!();

    // 2. Config demonstration
    println!("2. Configuration Manager");
    println!("--------------------------------------------------");
    let mut config = Config::new();
    config.set("opencog.version".to_string(), "1.0.0".to_string());
    config.set("atomspace.enabled".to_string(), "true".to_string());
    config.set("cogserver.port".to_string(), "17001".to_string());
    
    logger.info("Configuration loaded:");
    config.dump();
    println!();
    
    logger.info(&format!("Port setting: {}", config.get_or("cogserver.port", "unknown")));
    println!();

    // 3. Timer demonstration
    println!("3. Timer Utility");
    println!("--------------------------------------------------");
    let mut timer = Timer::new("Processing");
    timer.start();
    logger.info("Simulating some work...");
    
    // Simulate work
    let sum: u64 = (0..1_000_000).sum();
    let _ = sum;
    
    timer.stop();
    logger.info(&format!("Timer elapsed: {:.6}s", timer.elapsed_secs()));
    println!();

    // 4. Performance measurement
    println!("4. Performance Measurement");
    println!("--------------------------------------------------");
    measure_performance("Data processing", || {
        std::thread::sleep(Duration::from_millis(10));
    });
    println!();

    // 5. String utilities
    println!("5. String Utilities");
    println!("--------------------------------------------------");
    logger.info("String utilities demonstration:");
    let text = "OpenCog,AtomSpace,CogServer,Cogutil";
    let parts = StringUtils::split(text, ',');
    
    logger.info("Split result:");
    for part in &parts {
        println!("  - {}", part);
    }
    
    let joined = StringUtils::join(&parts, " + ");
    logger.info(&format!("Joined: {}", joined));
    
    logger.info(&format!("Uppercase: {}", StringUtils::to_upper("opencog rocks")));
    logger.info(&format!("Lowercase: {}", StringUtils::to_lower("OPENCOG ROCKS")));
    logger.info(&format!("Trimmed: '{}'", StringUtils::trim("  spaced out  ")));
    
    // Rust-specific: Case conversions
    logger.info(&format!("camelCase → snake_case: {}", 
                        StringUtils::camel_to_snake("myVariableName")));
    logger.info(&format!("snake_case → camelCase: {}", 
                        StringUtils::snake_to_camel("my_variable_name")));
    println!();

    // 6. Timing decorator
    println!("6. Function Timing");
    println!("--------------------------------------------------");
    
    let result = timeit("fibonacci", || {
        fn fibonacci(n: u32) -> u64 {
            if n <= 1 {
                n as u64
            } else {
                fibonacci(n - 1) + fibonacci(n - 2)
            }
        }
        fibonacci(30)
    });
    logger.info(&format!("Fibonacci(30) = {}", result));
    println!();

    // 7. Iterator demonstration (Rust-specific)
    println!("7. Iterators (Rust-specific)");
    println!("--------------------------------------------------");
    let squares: Vec<u32> = (0..10).map(|x| x * x).collect();
    logger.info(&format!("Squares: {:?}", squares));
    
    let evens: Vec<u32> = (0..20).filter(|x| x % 2 == 0).collect();
    logger.info(&format!("Even numbers: {:?}", evens));
    println!();

    // 8. Ownership demonstration
    println!("8. Ownership & Borrowing (Rust-specific)");
    println!("--------------------------------------------------");
    let s1 = String::from("OpenCog");
    let s2 = s1.clone(); // Explicit clone to avoid move
    logger.info(&format!("Original: {}, Clone: {}", s1, s2));
    
    // Borrowing example
    let config_ref = &config;
    logger.info(&format!("Has port config: {}", config_ref.has("cogserver.port")));
    println!();

    // 9. Result type for error handling
    println!("9. Result Type (Error Handling)");
    println!("--------------------------------------------------");
    match config.load_from_file("nonexistent.conf") {
        Ok(_) => logger.info("Config loaded successfully"),
        Err(e) => logger.warn(&format!("Config load failed (expected): {}", e)),
    }
    println!();

    logger.info("Cogutil demonstration complete!");
    println!("======================================================================");
    println!("Rust strengths demonstrated:");
    println!("  ✓ Memory safety without garbage collection");
    println!("  ✓ Ownership and borrowing system");
    println!("  ✓ Zero-cost abstractions");
    println!("  ✓ Pattern matching (enums, Option, Result)");
    println!("  ✓ Trait-based polymorphism");
    println!("  ✓ Powerful iterators");
    println!("  ✓ Compile-time guarantees");
    println!("======================================================================");
}

fn main() {
    demonstrate_cogutil();
}
